### PR TITLE
MLPModel

### LINKED ISSUE #829: MLP model
### 🚀 Feature Request

Add `MLPModel` model for etna deep model family

### Proposal

* You can use [as ref](https://github.com/tinkoff-ai/etna/blob/master/etna/models/nn/rnn.py)
```python
class MLPNet(DeepBaseNet):
    def __init__(
        self,
        input_size: int,
        hidden_size: List[int],
        lr: float,
        loss: "torch.nn.Module",
        optimizer_params: Optional[dict],
    ) -> None:
   pass
```

* For MLP we dont need encoder part: so `encoder_lenght=0`
* Typical batch sample would have shape `[encoder_lenght, input_size]`

### Test cases

* add unit tests for `make_sample`
* add seasonal overfit test
* add test for `step` and retrun values
* add test for `forward` and return values

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by DBcreator**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #829

### CODE REVIEW

File: tests/test_models/nn/test_mlp.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    assert first_sample["decoder_target"].shape == (decoder_length, 1)
+    np.testing.assert_equal(example_df[["target"]].iloc[:decoder_length], first_sample["decoder_target"])
+    np.testing.assert_equal(example_df[["target"]].iloc[1 : decoder_length + 1], second_sample["decoder_target"])
+
+
+def test_mlp_step():
+    torch.manual_seed(42)
+    model = MLPNet(input_size=3, hidden_size=[1], lr=1e-2, loss=None, optimizer_params=None)
+    batch = {"decoder_real": torch.Tensor([1, 2, 3]), "decoder_target": torch.Tensor([1, 2, 3]), "segment": "A"}
+    loss, decoder_target, _ = model.step(batch)
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:60]: This test is not very helpful I guess.
It's better to check contracts and method calls not some random magic values.

For example, you can check type of model.step output and mehod called inside step.

File: tests/test_models/nn/test_mlp.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    assert round(float(loss.detach().numpy()), 2) == 5.21
+    assert torch.all(decoder_target == torch.Tensor([1, 2, 3]))
+
+
+def test_mlp_forward():
+    torch.manual_seed(42)
+    model = MLPNet(input_size=3, hidden_size=[1], lr=1e-2, loss=None, optimizer_params=None)
+    batch = {"decoder_real": torch.Tensor([1, 2, 3]), "decoder_target": torch.Tensor([1, 2, 3]), "segment": "A"}
+    output = model.forward(batch)
+    assert round(float(output.detach().numpy()), 2) == -0.13
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:70]: The same problem here. We check `pytorch` not `etna` here.

At least you should check that all hidden layers were used and were called with proper inputs for example

File: tests/test_models/nn/test_mlp.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    decoder_length = 4
+
+    ts_samples = list(
+        MLPNet.make_samples(mlp_module, df=example_df, encoder_length=encoder_length, decoder_length=decoder_length)
+    )
+    first_sample = ts_samples[0]
+    second_sample = ts_samples[1]
+
+    assert first_sample["segment"] == "segment_1"
+    assert first_sample["decoder_real"].shape == (decoder_length, 0)
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:50]: Some strange case. I guess we can't use MLP with such shape at all

File: etna/models/nn/mlp.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        while True:
+            batch = _make(
+                df=df,
+                start_idx=start_idx,
+                decoder_length=decoder_length,
+            )
+            if batch is None:
+                break
+            yield batch
+            start_idx += 1
```
Comment (martins0n) [inline etna/models/nn/mlp.py:137]: We should increment by decoder_lenght. For now every point sampled `decoder_lenght` times

**Review by martins0n — CHANGES_REQUESTED** [review]

File: tests/test_models/nn/test_mlp.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    std = StandardScalerTransform(in_column="target")
+    ts_train.fit_transform([std, lag, fourier])
+
+    decoder_length = 14
+    model = MLPModel(
+        input_size=10,
+        encoder_length=0,
+        hidden_size=[10, 10, 10, 10, 10],
+        decoder_length=decoder_length,
+        trainer_params=dict(max_epochs=1000),
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:32]: mm, you can increase learning rate I guess

File: tests/test_models/nn/test_mlp.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    ts_train, ts_test = ts_dataset_weekly_function_with_horizon(horizon)
+    lag = LagTransform(in_column="target", lags=list(range(horizon, horizon + 4)))
+    fourier = FourierTransform(period=7, order=3)
+    std = StandardScalerTransform(in_column="target")
+    ts_train.fit_transform([std, lag, fourier])
+
+    decoder_length = 14
+    model = MLPModel(
+        input_size=10,
+        encoder_length=0,
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:29]: Encoder length is already the same

File: tests/test_models/nn/test_mlp.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    assert first_sample["decoder_target"].shape == (decoder_length, 1)
+    np.testing.assert_equal(example_df_with_lag[["target"]].iloc[:decoder_length], first_sample["decoder_target"])
+    np.testing.assert_equal(
+        example_df_with_lag[["target"]].iloc[decoder_length : 2 * decoder_length], second_sample["decoder_target"]
+    )
+
+
+def test_mlp_step():
+    model = MLPNet(input_size=3, hidden_size=[1], lr=1e-2, loss=None, optimizer_params=None)
+    batch = {"decoder_real": torch.Tensor([1, 2, 3]), "decoder_target": torch.Tensor([1, 2, 3]), "segment": "A"}
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:82]: Hm, I guess we have 2D parameters in decoder_real

File: tests/test_models/nn/test_mlp.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    future = ts_train.make_future(decoder_length)
+    model.fit(ts_train)
+    future = model.forecast(future, horizon=horizon)
+
+    mae = MAE("macro")
+    assert mae(ts_test, future) < 0.05
+
+
+@pytest.fixture()
+def example_df_with_lag(random_seed):
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:43]: Can you move this fixture to the beginning?
You don't need random_seed

File: tests/test_models/nn/test_mlp.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    )
+    future = ts_train.make_future(decoder_length)
+    model.fit(ts_train)
+    future = model.forecast(future, horizon=horizon)
+
+    mae = MAE("macro")
+    assert mae(ts_test, future) < 0.05
+
+
+@pytest.fixture()
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:42]: Do we really need this fixture? I guess you can use any other from general conftest.py

**Review by martins0n — CHANGES_REQUESTED** [review]

File: etna/models/nn/mlp.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
+from typing import Any
+from typing import Dict
+from typing import Iterable
+from typing import List
+from typing import Optional
+from typing import TypedDict
```
Comment (martins0n) [inline etna/models/nn/mlp.py:6]: There is some issues with `TypedDict` - it doesn't exists for python 3.7.

We should import it from typing_extensions

**Review by martins0n — CHANGES_REQUESTED** [review]

File: etna/models/nn/mlp.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        loss: "torch.nn.Module",
+        optimizer_params: Optional[dict],
+    ) -> None:
+        """Init MLP model.
+
+        Parameters
+        ----------
+        input_size:
+            size of the input feature space: target plus extra features
+        num_layers:
```
Comment (martins0n) [inline etna/models/nn/mlp.py:47]: There is no such param

File: etna/models/nn/mlp.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+        self.loss = nn.MSELoss() if loss is None else loss
+        self.optimizer_params = {} if optimizer_params is None else optimizer_params
+        layers = [nn.Linear(in_features=input_size, out_features=hidden_size[0]), nn.ReLU()]
+        for i in range(1, len(hidden_size)):
+            layers.append(nn.Linear(in_features=hidden_size[i - 1], out_features=hidden_size[i]))
+            layers.append(nn.ReLU())
+        layers.append(nn.Linear(in_features=hidden_size[-1], out_features=1))
+        self.mlp = nn.Sequential(*layers)
+
+    def forward(self, batch):
```
Comment (martins0n) [inline etna/models/nn/mlp.py:71]: we can add typing I guess

File: etna/models/nn/mlp.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+            val_dataloader_params=val_dataloader_params,
+            trainer_params=trainer_params,
+            split_params=split_params,
+        )
+        """Init MLP model.
+        Parameters
+        ----------
+        input_size:
+            size of the input feature space: target plus extra features
+        encoder_length:
```
Comment (martins0n) [inline etna/models/nn/mlp.py:193]: we should change order

**Review by martins0n — CHANGES_REQUESTED** [review]

**Review by martins0n — APPROVED** [review]
👍
